### PR TITLE
Bullet collision compound from hierarchy of meshes (Issue #301 fix part 2/2)

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

Now that the mesh hierarchy is stored (PR #309) , this PR exposes necessary assets and enables `BulletRigidObject` to construct a [btCompoundShape ](https://pybullet.org/Bullet/BulletFull/classbtCompoundShape.html) from a number of [btConvexHullShapes](https://pybullet.org/Bullet/BulletFull/classbtConvexHullShape.html) (for `RigidObjectType::OBJECT`) or [btBvhTriangleMeshShape](https://pybullet.org/Bullet/BulletFull/classbtBvhTriangleMeshShape.html) (for `RigidObjectType::SCENE`) with transforms accumulated from the stored hierarchy.

Mesh shifting, the previous method of aligning single mesh object origins with center of mass (COM) has been removed and replaced with a `SceneNode` shifting approach which generalizes to mesh hierarchies.

Note that these changes do break some previous assumptions and an update to the provided test object dataset `http://dl.fbaipublicfiles.com/habitat/objects_v0.1.zip` will be required upon merge. Specifically, `.glb` objects are now loaded with their global frame transforms properly (this was a hidden bug before), so collision shapes exported before as '.obj' files for `cheezit.glb` and `chefcan.glb` match the erroneously un-rotated meshes. Newly exported '.glb' versions with matching local_to_world rotation frames must be uploaded to remedy this.

## How Has This Been Tested

Local testing with Bullet debugDraw:

### Before (Issue #301 ):
bullet_hierarchal_mesh_bug

### After:
fixed_bullet_heirarchies

### Before refactoring mesh shift to play nice with hierarchal meshes we saw strange physics behavior for these objects:
heirarchal_COM_bug

### After:
chair_drop_bugfix

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/physics/bullet/BulletRigidObject.h [hunk 1]
```diff
@@ -62,15 +64,34 @@ class BulletRigidObject : public RigidObject {
...
+   * in a flat manner for efficiency.
+   * @param bCompound The @ref btCompoundShape being constructed.
+   * @param T The cumulative local-to-world transformation matrix constructed by
+   * composition down the @ref MeshTransformNode tree to the current node.
+   * @param meshGroup Access structure for collision mesh data.
+   * @param node The current @ref MeshTransformNode in the recursion.
+   */
+  void constructBulletConvexCompoundFromMeshes(
+      std::unique_ptr<btCompoundShape>& bCompound,
+      Magnum::Matrix4& T,
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.h]: Same request about matrix naming :)

Comment (erikwijmans) [reply src/esp/physics/bullet/BulletRigidObject.h]: Also, should this be a const reference?  If not, this should be a return value.

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 2]
```diff
@@ -103,11 +103,41 @@ bool BulletRigidObject::initializeScene(
...
 
+// recursively create the convex mesh shapes and add them to the compound in a
+// flat manner by accumulating transformations down the tree
+void BulletRigidObject::constructBulletConvexCompoundFromMeshes(
+    std::unique_ptr<btCompoundShape>& bCompound,
+    Magnum::Matrix4& T,
+    const std::vector<assets::CollisionMeshData>& meshGroup,
+    const assets::MeshTransformNode& node) {
+  Magnum::Matrix4 cT = T * node.T_parent_local;
+  btTransform t(cT);
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.cpp]:
```suggestion
```

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 3]
```diff
@@ -103,11 +103,41 @@ bool BulletRigidObject::initializeScene(
...
 }  // end BulletRigidObject::initializeScene
 
+// recursively create the convex mesh shapes and add them to the compound in a
+// flat manner by accumulating transformations down the tree
+void BulletRigidObject::constructBulletConvexCompoundFromMeshes(
+    std::unique_ptr<btCompoundShape>& bCompound,
+    Magnum::Matrix4& T,
+    const std::vector<assets::CollisionMeshData>& meshGroup,
+    const assets::MeshTransformNode& node) {
+  Magnum::Matrix4 cT = T * node.T_parent_local;
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.cpp]: Assuming that `T` is `T_world_parent` (seems like it is from the doc string), then `cT` is `T_world_local`?

File: src/esp/physics/bullet/BulletRigidObject.h [hunk 4]
```diff
@@ -62,15 +64,34 @@ class BulletRigidObject : public RigidObject {
...
+   * @param meshGroup The collision mesh data for the object.
    * @return true if initialized successfully, false otherwise.
    */
   bool initializeObject(
       const assets::PhysicsObjectAttributes& physicsObjectAttributes,
+      std::shared_ptr<btDiscreteDynamicsWorld> bWorld,
+      const assets::MeshMetaData& metaData,
+      const std::vector<assets::CollisionMeshData>& meshGroup);
+
+  /**
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.h]: Yay!  Thank you for the doc string.

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 5]
```diff
@@ -126,36 +156,10 @@ bool BulletRigidObject::initializeObject(
   //! Iterate through all mesh components for one object
   //! The components are combined into a convex compound shape
   bObjectShape_ = std::make_unique<btCompoundShape>();
-  for (const assets::CollisionMeshData& meshData : meshGroup) {
-    Corrade::Containers::ArrayView<Magnum::Vector3> v_data = meshData.positions;
-    Corrade::Containers::ArrayView<Magnum::UnsignedInt> ui_data =
-        meshData.indices;
+  Magnum::Matrix4 T;  // TODO: this should translate to the COM
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.cpp]: Doesn't look like this is used elsewhere, so deleted and do
```
constructBulletConvexCompoundFromMeshes(bObjectShape_, Magnum::Matrix4{Magnum::Math::IdentityInit}, meshGroup,
metaData.root);
```

Looks like Magnum does the right thing, (thank you @mosra) and `Magnum::Matrix4{}` would suffice :)

File: src/esp/physics/bullet/BulletRigidObject.cpp [hunk 6]
```diff
@@ -103,11 +103,41 @@ bool BulletRigidObject::initializeScene(
...
+
+    bObjectConvexShapes_.emplace_back(std::make_unique<btConvexHullShape>(
+        static_cast<const btScalar*>(mesh.positions.data()->data()),
+        mesh.positions.size(), sizeof(Magnum::Vector3)));
+
+    // TODO: need this for child shapes?
+    // bObjectConvexShapes_.back()->setMargin(margin);
+
+    //! Add to compound shape stucture
+    bObjectShape_->addChildShape(t, bObjectConvexShapes_.back().get());
```
Comment (erikwijmans) [inline src/esp/physics/bullet/BulletRigidObject.cpp]:
```suggestion
    bObjectShape_->addChildShape(btTransform{<whatever the name is>}, bObjectConvexShapes_.back().get());
```

File: src/esp/assets/ResourceManager.cpp [hunk 7]
```diff
@@ -745,6 +747,10 @@ bool ResourceManager::loadPTexMeshData(const AssetInfo& info,
         scene::SceneNode& node = parent->createChild();
         const quatf transform = info.frame.rotationFrameToWorld();
         node.setRotation(Magnum::Quaternion(transform));
+        Magnum::Matrix4 R = Magnum::Matrix4::from(
+            Magnum::Quaternion(transform).toMatrix(), Magnum::Vector3());
+        resourceDict_[filename].root.T_parent_local =
+            R * resourceDict_[filename].root.T_parent_local;
```
Comment (aclegg3) [inline src/esp/assets/ResourceManager.cpp]: Is there is nicer way to construct this Transform from an Eigen `quatf`?

**Comment by aclegg3** [comment]
CI tests are failing on frame equality assertions for test_sensors. Now that I am storing the transformation for each component of the loaded file, I'm rolling the `info.frame.rotationFrameToWorld()` transform into the root of the hierarchy rather than its parent. This likely breaks the expected `render_state` used for comparison in CI since the transform will have moved down the tree. Open to thoughts here. I could save that transformation elsewhere like `MeshMetaData::T_world_local` and convert things to expect this variable. Thoughts?

File: src/esp/assets/ResourceManager.cpp [hunk 8]
```diff
@@ -886,7 +894,9 @@ bool ResourceManager::loadGeneralMeshData(
     }  // forceReload
 
     const quatf transform = info.frame.rotationFrameToWorld();
-    newNode.setRotation(Magnum::Quaternion(transform));
```
Comment (erikwijmans) [inline src/esp/assets/ResourceManager.cpp]: Why was this deleted.

File: src/esp/assets/ResourceManager.cpp [hunk 9]
```diff
@@ -745,6 +747,10 @@ bool ResourceManager::loadPTexMeshData(const AssetInfo& info,
         scene::SceneNode& node = parent->createChild();
         const quatf transform = info.frame.rotationFrameToWorld();
         node.setRotation(Magnum::Quaternion(transform));
```
Comment (erikwijmans) [inline src/esp/assets/ResourceManager.cpp]: But this was kept?

**Comment by erikwijmans** [comment]
@aclegg3 -- I think I see the bug.  The bug has to due to loading a scene, swapping scenes, and then reloading that scene.  When you reload a scene, all the assets get cached and the `R` will end up getting applied multiple times in `metaData.root.T_parent_local = R * metaData.root.T_parent_local;` 

If you move the `metaData.root.T_parent_local = R * metaData.root.T_parent_local;` step to only be done when the file isn't loaded, everything passes.

File: src/esp/assets/ResourceManager.cpp (hunk 9 above)
Comment (aclegg3) [reply src/esp/assets/ResourceManager.cpp]: These functions work slightly differently. `loadGeneralMeshData` makes use of `addComponent` to handle mesh hierarchies. `loadPTexMeshData` assumes all meshes are already transformed by all but the global rotation transform (and as written assumes only 1 mesh, `... MeshMetaData(index, index) ...` ). I've added the global transform to the solitary `MeshTransformNode` purely for consistency here, it is likely not to be sued until collision mesh construction.

**Comment by aclegg3** [comment]
I noticed issues with loading arbitrary .glb files as scenes (like the wayfair objects themselves). The fix fits nicely into the re-design of this PR using MeshTransformNode to build the collision objects. Once this was done, mesh shifting no longer applied correctly to objects, so rather than push unstable functionality for the sake of small PRs, I rolled the refactor for bounding box shift in also. I'll re-frame the PR description to reflect this.
